File: DekafTestDB/tests/core/CommonQueryRunnerTest.java
package org.jetbrains.dekaf.core;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.dekaf.CommonIntegrationCase;
import org.jetbrains.dekaf.sql.Rewriters;
import org.jetbrains.dekaf.sql.SqlQuery;
import org.junit.Before;
import org.junit.FixMethodOrder;
import org.junit.Test;
import org.junit.runners.MethodSorters;

import java.sql.ResultSetMetaData;
import java.sql.SQLException;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.jetbrains.dekaf.core.Layouts.*;



/**
 * @author Leonid Bushuev from JetBrains
 **/
@FixMethodOrder(MethodSorters.JVM)
public class CommonQueryRunnerTest extends CommonIntegrationCase {

  protected static boolean isOracle =
      DB.rdbms().code.equalsIgnoreCase("ORACLE");


  @Before
  public void setUp() throws Exception {
    DB.connect();
    TH.prepareX1();
    TH.prepareX1000();
    TH.prepareX1000000();
  }


  public static class PrimitiveNumbers {
    byte  B;
    short S;
    int   I;
    long  L;
  }

  public static class BoxedNumbers {
    Byte    B;
    Short   S;
    Integer I;
    Long    L;
  }


  @Test
  public void query_existence_0 () {
    String queryText = "select 1 from "+(isOracle ? "dual" : "X1")+" where 1 is null";
    SqlQuery<Boolean> q = new SqlQuery<Boolean>(queryText, Layouts.existence());
    final Boolean b = query(q);
    Assertions.assertThat(b).isNotNull()
              .isFalse();
  }

  @Test
  public void query_existence_1() {
    String queryText = "select 1 "+(isOracle ? "from dual" : "");
    SqlQuery<Boolean> q = new SqlQuery<Boolean>(queryText, Layouts.existence());
    final Boolean b = query(q);
    Assertions.assertThat(b).isNotNull()
              .isTrue();
  }

  @Test
  public void query_primitive_numbers_positive() {
    final String queryText =
        "select 127 as B, 32767 as S, 2147483647 as I, 9223372036854775807 as L from X1";
    final SqlQuery<PrimitiveNumbers> query =
        new SqlQuery<PrimitiveNumbers>(queryText, Layouts.rowOf(Layouts.structOf(PrimitiveNumbers.class)));
    PrimitiveNumbers pn = query(query);

    Assertions.assertThat(pn.B).isEqualTo((byte)127);
    Assertions.assertThat(pn.S).isEqualTo((short)32767);
    Assertions.assertThat(pn.I).isEqualTo(2147483647);
    Assertions.assertThat(pn.L).isEqualTo(9223372036854775807L);
  }

  @Test
  public void query_primitive_numbers_negative() {
    final String queryText =
        "select -128 as B, -32768 as S, -2147483648 as I, -9223372036854775808 as L from X1";
    final SqlQuery<PrimitiveNumbers> query =
        new SqlQuery<PrimitiveNumbers>(queryText, Layouts.rowOf(Layouts.structOf(PrimitiveNumbers.class)));
    PrimitiveNumbers pn = query(query);

    Assertions.assertThat(pn.B).isEqualTo((byte)-128);
    Assertions.assertThat(pn.S).isEqualTo((short)-32768);
    Assertions.assertThat(pn.I).isEqualTo(-2147483648);
    Assertions.assertThat(pn.L).isEqualTo(-9223372036854775808L);
  }

  @Test
  public void query_boxed_numbers_positive() {
    final String queryText =
        "select 127 as B, 32767 as S, 2147483647 as I, 9223372036854775807 as L from X1";
    final SqlQuery<BoxedNumbers> query =
        new SqlQuery<BoxedNumbers>(queryText, Layouts.rowOf(Layouts.structOf(BoxedNumbers.class)));
    BoxedNumbers bn = query(query);

    Assertions.assertThat(bn.B).isEqualTo((byte) 127);
    Assertions.assertThat(bn.S).isEqualTo((short)32767);
    Assertions.assertThat(bn.I).isEqualTo(2147483647);
    Assertions.assertThat(bn.L).isEqualTo(9223372036854775807L);
  }


  @Test
  public void query_raw_numbers() {
    final String queryText =
        "select 127 as B, 32767 as S, 2147483647 as I, 9223372036854775807 as L from X1";
    final SqlQuery<Object[]> query =
        new SqlQuery<Object[]>(queryText, Layouts.rowOf(Layouts.rawArray()));
    final Object[] numbers = query(query);

    Assertions.assertThat(numbers).hasSize(4);
    Assertions.assertThat(numbers[0]).isInstanceOf(Number.class);
    Assertions.assertThat(numbers[1]).isInstanceOf(Number.class);
    Assertions.assertThat(numbers[2]).isInstanceOf(Number.class);
    Assertions.assertThat(numbers[3]).isInstanceOf(Number.class);
    Assertions.assertThat(((Number)numbers[0]).intValue()).isEqualTo(127);
    Assertions.assertThat(((Number)numbers[1]).intValue()).isEqualTo(32767);
    Assertions.assertThat(((Number)numbers[2]).intValue()).isEqualTo(2147483647);
    Assertions.assertThat(((Number)numbers[3]).longValue()).isEqualTo(9223372036854775807L);
  }

  @Test
  public void query_raw_strings() {
    final String queryText =
        "select 'C', 'String' from X1";
    final SqlQuery<Object[]> query =
        new SqlQuery<Object[]>(queryText, Layouts.rowOf(Layouts.rawArray()));
    final Object[] strings = query(query);

    Assertions.assertThat(strings).hasSize(2);

    Assertions.assertThat(strings[0]).isInstanceOf(String.class);
    Assertions.assertThat(strings[1]).isInstanceOf(String.class);

    Assertions.assertThat(strings[0]).isEqualTo("C");
    Assertions.assertThat(strings[1]).isEqualTo("String");
  }


  protected static class CalendarValues {
    public java.util.Date      javaDate;
    public java.sql.Date       sqlDate;
    public java.sql.Timestamp  sqlTimestamp;
    public java.sql.Time       sqlTime;
  }


  @Test
  public void query_calendar_values_now() {
    String queryText =
        "select NOW as javaDate, NOW as sqlDate, NOW as sqlTimestamp, NOW as sqlTime";
    if (isOracle) queryText += " from dual";
    SqlQuery<CalendarValues> query =
        new SqlQuery<CalendarValues>(queryText, Layouts.rowOf(Layouts.structOf(CalendarValues.class)))
          .rewrite(Rewriters.replace("NOW", sqlNow()));

    CalendarValues cv = query(query);

    Assertions.assertThat(cv.javaDate).isExactlyInstanceOf(java.util.Date.class);
    Assertions.assertThat(cv.sqlDate).isExactlyInstanceOf(java.sql.Date.class);
    Assertions.assertThat(cv.sqlTimestamp).isExactlyInstanceOf(java.sql.Timestamp.class);
    Assertions.assertThat(cv.sqlTime).isExactlyInstanceOf(java.sql.Time.class);
  }

  @Test
  public void query_calendar_values_parameters() {
    String queryText =
        queryCalendarValuesFromParameters();
    if (isOracle) queryText += " from dual";
    SqlQuery<CalendarValues> query =
        new SqlQuery<CalendarValues>(queryText, Layouts.rowOf(Layouts.structOf(CalendarValues.class)));

    CalendarValues cv = query(query,
                              new java.sql.Timestamp(System.currentTimeMillis()),
                              new java.sql.Timestamp(System.currentTimeMillis()),
                              new java.sql.Timestamp(System.currentTimeMillis()),
                              new java.sql.Timestamp(System.currentTimeMillis())/*,
                              new java.sql.Time(System.currentTimeMillis())*/
    );

    Assertions.assertThat(cv.javaDate).isExactlyInstanceOf(java.util.Date.class);
    Assertions.assertThat(cv.sqlDate).isExactlyInstanceOf(java.sql.Date.class);
    Assertions.assertThat(cv.sqlTimestamp).isExactlyInstanceOf(java.sql.Timestamp.class);
    Assertions.assertThat(cv.sqlTime).isExactlyInstanceOf(java.sql.Time.class);
  }

  @NotNull
  protected String queryCalendarValuesFromParameters() {
    return "select ? as javaDate, ? as sqlDate, ? as sqlTimestamp, ? as sqlTime";
  }

  @NotNull
  protected String sqlNow() {
    //noinspection SpellCheckingInspection
    return "current_timestamp";
  }


  @Test
  public void query_1000_values() {
    List<Integer> values =
        DB.inTransaction(tran ->
            tran.query("select X from X1000 order by 1", Layouts.listOf(Layouts.oneOf(Integer.class))).run()
        );
    Assertions.assertThat(values).isNotNull()
              .hasSize(1000)
              .contains(1,2,3,4,998,999,1000);
  }


  @Test
  public void query_1000000_values() {
    List<Integer> values =
        DB.inTransaction(tran -> tran.query("select X from X1000000 order by 1", Layouts.listOf(
                Layouts.oneOf(Integer.class))).run());
    Assertions.assertThat(values).isNotNull()
              .hasSize(1000000)
              .contains(1,2,3,4,999998,999999,1000000);
  }


  @Test
  public void access_metadata() {
    final SqlQuery<List<Number>> query =
        new SqlQuery<List<Number>>("select X from X1000", Layouts.listOf(Layouts.oneOf(Number.class)));

    DB.inSessionDo(session -> {

        DBQueryRunner<List<Number>> qr = session.query(query).packBy(10);
        qr.run();
        ResultSetMetaData md =
            qr.getSpecificService(ResultSetMetaData.class,
                                  ImplementationAccessibleService.Names.JDBC_METADATA);
        Assertions.assertThat(md).isNotNull();

        String columnName = null;
        try {
          columnName = md.getColumnName(1);
        }
        catch (SQLException e) {
          throw new RuntimeException(e.getMessage(), e);
        }
        Assertions.assertThat(columnName).isEqualToIgnoringCase("X");

    });
  }


  protected <T> T query(@NotNull final SqlQuery<T> query) {
    return DB.inTransaction(tran -> {
        return tran.query(query).run();
    });
  }

  protected <T> T query(@NotNull final SqlQuery<T> query, final Object... params) {
    return DB.inTransaction(tran -> tran.query(query).withParams(params).run());
  }

}

File: DekafTestDB/tests/core/CommonQueryRunnerTest.kt
@file:Suppress("platform_class_mapped_to_kotlin")

package org.jetbrains.dekaf.core

import org.assertj.core.api.Assertions
import org.jetbrains.dekaf.CommonIntegrationCase
import org.jetbrains.dekaf.Oracle
import org.jetbrains.dekaf.assertions.IsNotNull
import org.jetbrains.dekaf.assertions.expected
import org.jetbrains.dekaf.sql.SqlQuery
import org.jetbrains.dekaf.text.Rewriters
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.lang.Number
import java.sql.ResultSetMetaData
import java.sql.SQLException
import kotlin.Any
import kotlin.Long
import kotlin.RuntimeException
import kotlin.String
import kotlin.Suppress


/**
 * @author Leonid Bushuev
 */
class CommonQueryRunnerTest : CommonIntegrationCase() {


    companion object {

        @BeforeAll @JvmStatic
        fun setUp() {
            DB.connect()
            TH.prepareX1()
            TH.prepareX1000()
            TH.prepareX1000000()
        }

        protected var isOracle = DB.rdbms() == Oracle.RDBMS

    }


    @Test
    fun query_existence_0() {
        val queryText = "select 1 from " + (if (isOracle) "dual" else "X1") + " where 1 is null"
        val q = SqlQuery(queryText, layoutExistence())
        val b = query(q)
        Assertions.assertThat(b).isNotNull
                .isFalse
    }

    @Test
    fun query_existence_1() {
        val queryText = "select 1 " + if (isOracle) "from dual" else ""
        val q = SqlQuery(queryText, layoutExistence())
        val b = query(q)
        Assertions.assertThat(b).isNotNull
                .isTrue
    }

    @Test
    fun query_primitive_numbers_basic() {
        val queryText = "select 1 as B, 2 as S, 3 as I, 4 as L"
        val layout = layoutOneRowOf(rowStructOf(PrimitiveNumbers::class.java))
        val query = SqlQuery(queryText, layout)
        val pn = query(query)

        pn!!

        Assertions.assertThat(pn.B).isEqualTo(1.toByte())
        Assertions.assertThat(pn.S).isEqualTo(2.toShort())
        Assertions.assertThat(pn.I).isEqualTo(3)
        Assertions.assertThat(pn.L).isEqualTo(4L)
    }

    @Test
    fun query_primitive_numbers_positive() {
        val queryText = "select 127 as B, 32767 as S, 2147483647 as I, 9223372036854775807 as L"
        val query = SqlQuery(queryText, layoutOneRowOf(rowStructOf(PrimitiveNumbers::class.java)))
        val pn = query(query)

        pn!!

        Assertions.assertThat(pn.B).isEqualTo(127.toByte())
        Assertions.assertThat(pn.S).isEqualTo(32767.toShort())
        Assertions.assertThat(pn.I).isEqualTo(2147483647)
        Assertions.assertThat(pn.L).isEqualTo(9223372036854775807L)
    }

    @Test
    fun query_primitive_numbers_negative() {
        val queryText = "select -128 as B, -32768 as S, -2147483648 as I, -9223372036854775808 as L"
        val query = SqlQuery(queryText, layoutOneRowOf(rowStructOf(PrimitiveNumbers::class.java)))
        val pn = query(query)

        Assertions.assertThat(pn!!.B).isEqualTo((-128).toByte())
        Assertions.assertThat(pn.S).isEqualTo((-32768).toShort())
        Assertions.assertThat(pn.I).isEqualTo(-2147483648)
        Assertions.assertThat(pn.L).isEqualTo(Long.MIN_VALUE)   // hello,  KT-17172
    }

    @Test
    fun query_boxed_numbers_positive() {
        val queryText = "select 127 as B, 32767 as S, 2147483647 as I, 9223372036854775807 as L from X1"
        val query = SqlQuery(queryText, layoutOneRowOf(rowStructOf(BoxedNumbers::class.java)))
        val bn = query(query)

        Assertions.assertThat(bn!!.B).isEqualTo(127.toByte())
        Assertions.assertThat(bn.S).isEqualTo(32767.toShort())
        Assertions.assertThat(bn.I).isEqualTo(2147483647)
        Assertions.assertThat(bn.L).isEqualTo(9223372036854775807L)
    }


    @Test
    fun query_raw_numbers() {
        val queryText = "select 127 as B, 32767 as S, 2147483647 as I, 9223372036854775807 as L from X1"
        val query = SqlQuery(queryText, layoutOneRowOf(rowArrayOf<Number>()))
        val numbers = query(query)

        numbers!!

        Assertions.assertThat(numbers).hasSize(4)
        Assertions.assertThat(numbers[0]).isInstanceOf(Number::class.java)
        Assertions.assertThat(numbers[1]).isInstanceOf(Number::class.java)
        Assertions.assertThat(numbers[2]).isInstanceOf(Number::class.java)
        Assertions.assertThat(numbers[3]).isInstanceOf(Number::class.java)
        Assertions.assertThat((numbers[0] as Number).intValue()).isEqualTo(127)
        Assertions.assertThat((numbers[1] as Number).intValue()).isEqualTo(32767)
        Assertions.assertThat((numbers[2] as Number).intValue()).isEqualTo(2147483647)
        Assertions.assertThat((numbers[3] as Number).longValue()).isEqualTo(9223372036854775807L)
    }

    @Test
    fun query_raw_strings() {
        val queryText = "select 'C', 'String' from X1"
        val query = SqlQuery(queryText, layoutOneRowOf(rowArrayOf<Any>()))
        val strings = query(query)

        strings!!

        Assertions.assertThat(strings).hasSize(2)

        Assertions.assertThat(strings[0]).isInstanceOf(String::class.java)
        Assertions.assertThat(strings[1]).isInstanceOf(String::class.java)

        Assertions.assertThat(strings[0]).isEqualTo("C")
        Assertions.assertThat(strings[1]).isEqualTo("String")
    }


    @Test
    fun query_calendar_values_now() {
        var queryText = "select NOW as javaDate, NOW as sqlDate, NOW as sqlTimestamp, NOW as sqlTime"
        if (isOracle) queryText += " from dual"
        val query = SqlQuery(queryText, layoutOneRowOf(rowStructOf(CalendarValues::class.java)))
                .rewrite(Rewriters.replace("NOW", sqlNow()))

        val cv = query(query)

        Assertions.assertThat(cv!!.javaDate).isExactlyInstanceOf(java.util.Date::class.java)
        Assertions.assertThat(cv.sqlDate).isExactlyInstanceOf(java.sql.Date::class.java)
        Assertions.assertThat(cv.sqlTimestamp).isExactlyInstanceOf(java.sql.Timestamp::class.java)
        Assertions.assertThat(cv.sqlTime).isExactlyInstanceOf(java.sql.Time::class.java)
    }

    @Test
    fun query_calendar_values_parameters() {
        var queryText = queryCalendarValuesFromParameters()
        if (isOracle) queryText += " from dual"
        val query = SqlQuery(queryText, layoutOneRowOf(rowStructOf(CalendarValues::class.java)))

        val cv = query(query,
                       java.sql.Timestamp(System.currentTimeMillis()),
                       java.sql.Timestamp(System.currentTimeMillis()),
                       java.sql.Timestamp(System.currentTimeMillis()),
                       java.sql.Timestamp(System.currentTimeMillis())
        )

        Assertions.assertThat(cv.javaDate).isExactlyInstanceOf(java.util.Date::class.java)
        Assertions.assertThat(cv.sqlDate).isExactlyInstanceOf(java.sql.Date::class.java)
        Assertions.assertThat(cv.sqlTimestamp).isExactlyInstanceOf(java.sql.Timestamp::class.java)
        Assertions.assertThat(cv.sqlTime).isExactlyInstanceOf(java.sql.Time::class.java)
    }


    protected fun queryCalendarValuesFromParameters(): String {
        return "select ? as javaDate, ? as sqlDate, ? as sqlTimestamp, ? as sqlTime"
    }

    protected fun sqlNow(): String {

        return "current_timestamp"
    }


    @Test
    fun query_1000_values() {
        val values = CommonIntegrationCase.DB.inTransaction<List<Integer>> { tran -> tran.query("select X from X1000 order by 1", layoutListOf(rowValueOf(Integer::class.java))).run() }
        Assertions.assertThat(values).isNotNull
                .hasSize(1000)
                .contains(Integer(1), Integer(2), Integer(3), Integer(998), Integer(999), Integer(1000))
    }


    @Test
    fun query_1000000_values() {
        val values = CommonIntegrationCase.DB.inTransaction<List<Integer>> { tran -> tran.query("select X from X1000000 order by 1", layoutListOf(rowValueOf(Integer::class.java))).run() }
        Assertions.assertThat(values).isNotNull
                .hasSize(1000000)
                .contains(Integer(1), Integer(2), Integer(3), Integer(999998), Integer(999999), Integer(1000000))
    }


    @Test
    fun access_metadata() {
        val query = SqlQuery("select 123456789 as X", layoutListOf(rowValueOf(Number::class.java)))

        CommonIntegrationCase.DB.inSessionDo { session ->

            val qr = session.query(query).packBy(10).execute()
            val md = qr.getSpecificService(ResultSetMetaData::class.java,
                                           ImplementationAccessibleService.Names.JDBC_METADATA)

            md expected IsNotNull
            md!!

            val columnName: String?
            try {
                columnName = md.getColumnName(1)
            }
            catch (e: SQLException) {
                throw RuntimeException(e.message, e)
            }

            Assertions.assertThat(columnName).isEqualToIgnoringCase("X")
        }
    }


    protected fun <T> query(query: SqlQuery<T>): T? {
        return CommonIntegrationCase.DB.inTransaction<T> { tran -> tran.query(query).run() }
    }

    protected fun <T> query(query: SqlQuery<T>, vararg params: Any): T {
        return CommonIntegrationCase.DB.inTransaction<T> { tran -> tran.query(query).withParams(*params).run() }
    }

}
